### PR TITLE
Update/adds supabase integration

### PR DESCRIPTION
**Pull request by doingandlearning**
This adds the Supabase integration into the project. It also adds the store and function to trigger the data fetch if it hasn't happened already.

To test, you'll need the Supabase key and url to add into the `.env` file - I'll share these over Slack.

1. Run `yarn install`
2. Navigate to the route page and confirm everything works as expected (nothing is broken)
3. Make a newline between the opening div and the SocialHead component in CommunityPage.vue and add `{{ JSON.stringify(supabase.topics) }}` or something similar. 
4. View on the frontend and confirm that the data is present.

In my testing branch, you can see how this would work with the Graph and List component (https://github.com/phivk/gradu.al/blob/temp/experimenting-with-forms/pages/viz.vue). There's a weird thing where spreading in a new array `[...array]` is still being treated as a Vuex object - yuck! So, for now I've got a horrible use of JSON.stringify and JSON.parse to make this work :)

### CODE REVIEW

**Review by phivk — COMMENTED** [review]
finally got to review this 😅 

It works! Initially I was getting CORS errors, which was confusing as actually the issue seems to be that the SupaBase project was paused. When I restored the project, everything worked.

I tried the suggested `{{ JSON.stringify(supabase.topics) }}` and it works.
I also tried `{{ supabase.topics }}`, which also works.

This make me ask two questions:

1. Are you sure the Stringify hack is necessary?
2. Will you fix it as part of this PR or do you prefer to save it for later? On first sight it seems like this is such a common usecase that we should be able to figure this out, but let me know your thoughts 😉

**Comment by doingandlearning** [comment]
> This make me ask two questions:
> 
> 1. Are you sure the Stringify hack is necessary?

The issue is that we're logging the whole of the object here which is why we are stringifying. Normally, we'd use a `v-for` or access a particular key. This is merely to demonstrate that the data is getting from Supabase to the page.

> 2. Will you fix it as part of this PR or do you prefer to save it for later? On first sight it seems like this is such a common usecase that we should be able to figure this out, but let me know your thoughts 😉

I think the above comment addresses this - this PR was about getting the data from Supabase to the app and the stringify was to demonstrate that is true. Going forward we don't need to stringify because we've accessing the actual data we want. 

Hope that answers your questions :)

**Comment by phivk** [comment]
@doingandlearning 
I've had a closer look and will leave some more detailed review comments.

On a general note, I would like to change the base brach of the three PRs to https://github.com/phivk/gradu.al/tree/supabase-integration so we can work on the styling, integration and any other further development separately from `develop`. I will update the base branches now.

**EDIT:** Another big question I have is about multiple communities. This current setup deals with storing data for a single community.

What is the idea for dealing with that? Will they each have their own Supabase project (doesn't seem ideal)? Will they be represented in different tables in one SB project (also not so sure)? 
I am sure there are solutions ready at hand in SupaBase land to deal with this use case.

File: store/supabase.js [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+    state.nodes = nodes;
+  },
+};
+
+function createNodes(members, topics, mappings) {
+  let nodes = [];
+  let memberNames = {};
+  for (const member of members) {
+    nodes.push({
+      _cssClass: "Member",
```
Comment (phivk) [inline store/supabase.js]: the createNodes function prepares the graph data structure specifically as expected by the `vue-d3-network` lib, and thus sets a lot of display attributes like `_cssClass` and `_svgAttrs`. 
I think this should not be the responsibility of the store and therefore this function could better be defined in an external utility file.

File: store/supabase.js [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+  },
+  setEdges(state, edges) {
+    state.edges = edges;
+  },
+  setNodes(state, nodes) {
+    state.nodes = nodes;
+  },
+};
+
+function createNodes(members, topics, mappings) {
```
Comment (phivk) [inline store/supabase.js]: this function create a graph, made up from Nodes and Edges, specifically targeting `vue-d3-network`. Could that be reflected in the function name? 
`createVueD3NetworkGraph()`?
`createNodesEdgesForVueD3Network()`?
...?

File: store/supabase.js [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+  async fetchData({ commit }) {
+    // hit supabase to get topics, members and mappings
+    const { data: topics, error: topicError } = await this.$supabase
+      .from("topics")
+      .select("*");
+    const { data: members, error: memberError } = await this.$supabase
+      .from("members")
+      .select("*");
+    const { data: mapping, error: mappingError } = await this.$supabase
+      .from("members_topics")
```
Comment (phivk) [inline store/supabase.js]: why call it `mapping` here and `members_topics` in SupaBase?
Let's make it consistent.

Comment (phivk) [reply store/supabase.js]: also `mapping` is singluar here, but `mappings` is plural in the `createNodes()` signature, which is confusing

File: store/supabase.js [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+      id: `topic-${topic.id}`,
+      sharerNames: [],
+      learnerNames: [],
+    });
+  }
+
+  const edges = [];
+  const counter = { learners: {}, sharers: {} };
+
+  for (const topic of mappings.filter((item) => item.learner)) {
```
Comment (phivk) [inline store/supabase.js]: It takes me quite some time to wrap my head around how this loop is adding learning relationships to the edges array. Perhaps add a little comment that describes what the function is doing?

File: store/supabase.js [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+      type: "learn",
+      _color: "#f1955b",
+      sid: `member-${topic.learner}`,
+      tid: `topic-${topic.topic}`,
+      _svgAttrs: { "stroke-width": "2", opacity: 0.5 },
+    });
+    counter.learners[`topic-${topic.topic}`] = counter.learners[
+      `topic-${topic.topic}`
+    ] || { count: 0, names: [] };
+    counter.learners[`topic-${topic.topic}`].count++;
```
Comment (phivk) [inline store/supabase.js]: we used to keep track of `sharerCount` and `learnerCount` in https://github.com/phivk/gradu.al/blob/1b34d7f61c10962ab7625b0f467305dd39db308c/fetchData.js#L137

I guess you are proposing a slight refactor of the nodes edges data structure?
Please note that `components/MostPopularSkillsSection.vue` is relying on `skill.learnerCount`
see https://github.com/phivk/gradu.al/blob/1b34d7f61c10962ab7625b0f467305dd39db308c/components/MostPopularSkillsSection.vue#L44

File: store/supabase.js [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+    counter.learners[`topic-${topic.topic}`].names.push(
+      memberNames[topic.learner]
+    );
+  }
+
+  for (const share of mappings.filter((item) => item.sharer)) {
```
Comment (phivk) [inline store/supabase.js]: maybe another comment here to describe what this function is doing?

File: store/supabase.js (hunk 5 above)
Comment (phivk) [reply store/supabase.js]: oh, now I see that you're mapping this below...
I guess it is efficient to do the counting while iterating over the edges, but it makes the function hard to parse for me as a human. Not sure about the best approach here.

File: store/supabase.js [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+    );
+  }
+
+  for (const [topic, detail] of Object.entries(counter.learners)) {
+    const indexToReplace = nodes.findIndex((el) => el.id === topic);
+    const topicToUpdate = nodes[indexToReplace];
+    topicToUpdate.learnerCount = detail.count;
+    topicToUpdate.learnerNames = detail.names;
+    nodes.splice(indexToReplace, 1);
+    nodes.push(topicToUpdate);
```
Comment (phivk) [inline store/supabase.js]: could this be simpler via
```
nodes[indexToReplace] = topicToUpdate
```
?

File: store/supabase.js [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+    counter.sharers[`topic-${share.topic}`] = counter.sharers[
+      `topic-${share.topic}`
+    ] || { count: 0, names: [] };
+    counter.sharers[`topic-${share.topic}`].count++;
+    counter.sharers[`topic-${share.topic}`].names.push(
+      memberNames[share.sharer]
+    );
+  }
+
+  for (const [topic, detail] of Object.entries(counter.learners)) {
```
Comment (phivk) [inline store/supabase.js]: another documentation comment would be nice here.

File: store/supabase.js [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+    const indexToReplace = nodes.findIndex((el) => el.id === topic);
+    const topicToUpdate = nodes[indexToReplace];
+    topicToUpdate.learnerCount = detail.count;
+    topicToUpdate.learnerNames = detail.names;
+    nodes.splice(indexToReplace, 1);
+    nodes.push(topicToUpdate);
+  }
+
+  for (const [topic, detail] of Object.entries(counter.sharers)) {
```
Comment (phivk) [inline store/supabase.js]: and a documentation comment here :)

**Review by phivk — CHANGES_REQUESTED** [review]

**Comment by doingandlearning** [comment]
I've updated this branch based on our previous discussions.

**Review by phivk — APPROVED** [review]
